### PR TITLE
fix: generate complete variant binaries properly

### LINKED ISSUE #27608: Complete variant for the elastic agent is not generating the right binaries
When running `mage package`, the binaries generated by the elastic-agent are incorrect:
- there are three variants: "default", ubi8 and complete.
- the docker image for the ubi8 variant is properly generated under build/distributions
- the docker image for the complete variant is generated under build/distributions, BUT overriding the default image.

I suspect that this has been caused by #27350 and #27052

### Steps to reproduce
1. Run `cd x-pack/elastic-agent`
2. Run `PLATFORMS="linux/amd64" mage package` (Narrowing down the process to build only one platform)
3. Wait for the process to finish... (_10 min?_)

### Expected behaviour
Three docker image binaries are generated, one per variant (default, ubi8 and complete), being the complete one the heaviest in weight (contains node runtime plus libs).
```shell
$ ls -l x-pack/elastic-agent/build/distributions | grep docker
-rw-r--r--  1 mdelapenya  staff  244359238 26 Aug 13:51 elastic-agent-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        175 26 Aug 13:51 elastic-agent-8.0.0-linux-amd64.docker.tar.gz.sha512
-rw-r--r--  1 mdelapenya  staff  614539725 26 Aug 13:55 elastic-agent-complete-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        184 26 Aug 13:56 elastic-agent-complete-8.0.0-linux-amd64.docker.tar.gz.sha512
-rw-r--r--  1 mdelapenya  staff  163226167 26 Aug 13:50 elastic-agent-ubi8-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        180 26 Aug 13:50 elastic-agent-ubi8-8.0.0-linux-amd64.docker.tar.gz.sha512
```

When you docker load the default image, the default image is loaded into the Docker engine:
```shell
$ docker load -i x-pack/elastic-agent/build/distributions/elastic-agent-8.0.0-linux-amd64.docker.tar.gz 
Loaded image: docker.elastic.co/beats/elastic-agent:8.0.0
```

### Current behaviour
Two docker image binaries are generated, one per variant (default and ubi8, only).
```shell
$ ls -l x-pack/elastic-agent/build/distributions | grep docker
-rw-r--r--   1 mdelapenya  staff  608875722 26 Aug 12:57 elastic-agent-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--   1 mdelapenya  staff        175 26 Aug 12:58 elastic-agent-8.0.0-linux-amd64.docker.tar.gz.sha512
-rw-r--r--   1 mdelapenya  staff  163226167 26 Aug 12:56 elastic-agent-ubi8-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--   1 mdelapenya  staff        180 26 Aug 12:56 elastic-agent-ubi8-8.0.0-linux-amd64.docker.tar.gz.sha512
```

When you docker load the default image, the complete image is loaded into the Docker engine:
```shell
$ docker load -i x-pack/elastic-agent/build/distributions/elastic-agent-8.0.0-linux-amd64.docker.tar.gz 
Loaded image: docker.elastic.co/beats/elastic-agent-complete:8.0.0
```

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
This PR is enriching the image name for the "complete" variant of the Elastic Agent when building it using `mage package`. It simply adds an extra spec that must be added to the complete variant. We are adding an additional build package for the Elastic Agent, exactly the same as for the UBI8 variant.

Given I'm not a Beats build expert, I'm copying what the ubi8 variant does, because it seems the `variants` keyword used in the past is not doing its job. Maybe @jsoriano or @kvch knows more about it.

## Why is it important?
We realised that CI builds were generating incorrect binaries for the agent, as the elastic-agent docker image (TAR.GZ binary) was pushed to the internal GCP bucket only once, as described in #27608. Then, any consumer of those binaries was always receiving a +500MB file when downloading it from the bucket. An example of a consumer is the e2e-testing framework, which gets a binary for a merge commit or a PR in Beats project, and runs the E2E tests against those non-released delta binaries. We noticed the download times for the default artifact increased up to the double, also missing the `complete` TAR.GZ binary in the GCP bucket.

With this fix we expect the CI artifacts, which are super useful for bisecting errors, will be generated properly again.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ] Is there any other mage package combination that I missed?
- [ ] Is the spec I generated complete?
- [ ] Is there a `variants` key in the packages.yml file? How is it used, because without this fix it seemed not to work as expected?

## How to test this PR locally

```shell
cd x-pack/elastic-agent
PLATFORMS="linux/amd64" mage package
```
After around 10 minutes... check for the binaries produced by the build system:
```shell
ls -l x-pack/elastic-agent/build/distributions | grep docker
-rw-r--r--  1 mdelapenya  staff  244359238 26 Aug 13:51 elastic-agent-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        175 26 Aug 13:51 elastic-agent-8.0.0-linux-amd64.docker.tar.gz.sha512
-rw-r--r--  1 mdelapenya  staff  614539725 26 Aug 13:55 elastic-agent-complete-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        184 26 Aug 13:56 elastic-agent-complete-8.0.0-linux-amd64.docker.tar.gz.sha512
-rw-r--r--  1 mdelapenya  staff  163226167 26 Aug 13:50 elastic-agent-ubi8-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        180 26 Aug 13:50 elastic-agent-ubi8-8.0.0-linux-amd64.docker.tar.gz.sha512
```

## Related issues

- Caused by #27350 and #27052
- Fixes #27608
- Relates #28134

### CODE REVIEW

**Comment by kvch** [comment]
What is up with this PR? Is it still required? Personally, I like this approach because it is consistent with the current packaging.

**Comment by mdelapenya** [comment]
> What is up with this PR? Is it still required? Personally, I like this approach because it is consistent with the current packaging.

Do you think of any incompatibility with #27621? As you suggest, I'd like to see more robustness in the build process too

**Comment by kvch** [comment]
No, I don't see any incompatibility.

**Comment by mdelapenya** [comment]
jenkins run the test please

**Comment by mdelapenya** [comment]
CI is passing. @kvch @jsoriano could you please give a review on this PR?

**Review by kvch — APPROVED** [review]
LGTM, but please wait for one more approval.

**Review by jsoriano — CHANGES_REQUESTED** [review]
With these changes the following images are generated, one of them containing two suffixes, and another one with a `-complete` suffix but seemingly having the same content as the one without this suffix. Is this expected? :thinking: 
```
REPOSITORY                                                TAG                        IMAGE ID       CREATED             SIZE
docker.elastic.co/beats/elastic-agent-complete-complete   8.0.0                      f5bc5bc3e5fd   34 seconds ago      1.55GB
docker.elastic.co/beats/elastic-agent                     8.0.0                      4069689b3330   3 minutes ago       523MB
docker.elastic.co/beats/elastic-agent-complete            8.0.0                      61d02e44734e   3 minutes ago       523MB
docker.elastic.co/beats/elastic-agent-ubi8                8.0.0                      51cecc7df1fb   3 minutes ago       318MB
```

**Comment by mdelapenya** [comment]
> With these changes the following images are generated, one of them containing two suffixes, and another one with a `-complete` suffix but seemingly having the same content as the one without this suffix. Is this expected? 🤔
> 

I think because #27621 was merged before this one, there are conflicts in resolving the issue described here. As @kvch said, these changes are affecting configuration, while that PR is updating the build system to cover this use case.

I tested my changes before #27621, I can give it a try again, with and without that commit.

**Comment by mdelapenya** [comment]
@jsoriano @kvch @andrewkroh @andrewvc I updated this PR and verified that:

```shell
cd x-pack/elastic-agent
PLATFORMS="linux/amd64" mage package
```
- the Docker engine contains 3 Docker images, _default_, complete and ubi8, each of them with the right size (complete is around 1'5Gb)
- generates the Docker binaries for each variant: the `build/distributions` directory contains 3 `tar.gz` files, one per variant (default, complete and ubi8), each of them with the right size (complete includes Node libs)

```shell
$ ls -l x-pack/elastic-agent/build/distributions/
total 2791688
-rw-r--r--@ 1 mdelapenya  staff  119939138  9 Sep 17:18 elastic-agent-8.0.0-amd64.deb
-rw-r--r--  1 mdelapenya  staff        159  9 Sep 17:18 elastic-agent-8.0.0-amd64.deb.sha512
-rw-r--r--  1 mdelapenya  staff  244353172  9 Sep 17:18 elastic-agent-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        175  9 Sep 17:18 elastic-agent-8.0.0-linux-amd64.docker.tar.gz.sha512
-rw-r--r--  1 mdelapenya  staff  120003980  9 Sep 17:17 elastic-agent-8.0.0-linux-x86_64.tar.gz
-rw-r--r--  1 mdelapenya  staff        169  9 Sep 17:17 elastic-agent-8.0.0-linux-x86_64.tar.gz.sha512
-rw-r--r--  1 mdelapenya  staff  119745407  9 Sep 17:18 elastic-agent-8.0.0-x86_64.rpm
-rw-r--r--  1 mdelapenya  staff        160  9 Sep 17:18 elastic-agent-8.0.0-x86_64.rpm.sha512
-rw-r--r--  1 mdelapenya  staff  615426504  9 Sep 17:23 elastic-agent-complete-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        184  9 Sep 17:24 elastic-agent-complete-8.0.0-linux-amd64.docker.tar.gz.sha512
-rw-r--r--  1 mdelapenya  staff  163243411  9 Sep 17:18 elastic-agent-ubi8-8.0.0-linux-amd64.docker.tar.gz
-rw-r--r--  1 mdelapenya  staff        180  9 Sep 17:18 elastic-agent-ubi8-8.0.0-linux-amd64.docker.tar.gz.sha512
```

- Docker loading each of these artifact results in having the Docker engine loading the right image:

```shell
$ docker images | grep agent | grep -v SNAP
docker.elastic.co/beats/elastic-agent-complete          8.0.0                                      7cc90f716192   40 minutes ago      1.55GB
docker.elastic.co/beats/elastic-agent-ubi8              8.0.0                                      58ed21edadcd   44 minutes ago      297MB
docker.elastic.co/beats/elastic-agent                   8.0.0                                      a9c9517dcace   44 minutes ago      521MB
```

TBH, I'm pretty sure I'm acting as a surgeon in their first to-open-heart surgery with this PR affecting the build system. I'd appreciate your review. Thanks!

File: x-pack/osquerybeat/beater/config_plugin_test.go
Comment (mdelapenya) [inline x-pack/osquerybeat/beater/config_plugin_test.go:19]: Not related, but found

File: dev-tools/mage/pkgtypes.go
Comment (jsoriano) [inline dev-tools/mage/pkgtypes.go:51]: If `Variant` is not needed anymore here, I guess that the rest of https://github.com/elastic/beats/pull/27621 can be reverted too?

File: dev-tools/mage/dockerbuilder.go
Comment (jsoriano) [inline dev-tools/mage/dockerbuilder.go:197]: `variant` not used anymore here, please remove the parameter from these methods.

**Review by jsoriano — CHANGES_REQUESTED** [review]
Looks good, I like that variants are more just another image in the yml file and less a special case in `dockerbuilder.go`.

Added some comments about removing unneeded code and fully reverting https://github.com/elastic/beats/pull/27621. I wonder if we can even remove more code for variants in the docker builder.

File: dev-tools/mage/dockerbuilder.go
Comment (mdelapenya) [reply dev-tools/mage/dockerbuilder.go:197]: I tried that, but then the loop for the existing variants does not make sense. As I'm not a domain expert, not sure if it's connected to the Variants attribute in the types struct, which is used at the Dockerfile template for the agent, and finally connected to the `variants` array in the YAML file... 😱

Comment (jsoriano) [reply dev-tools/mage/dockerbuilder.go:197]: Well, you can leave in other places by now, removing it probably requires further refactors, but at least in this method it is not used anymore.
```suggestion
func (b *dockerBuilder) dockerBuild() (string, error) {
```

**Comment by mdelapenya** [comment]
Once https://github.com/elastic/beats/pull/28134 is merged, we should update this PR and possibly resolve conflicts

**Review by cachedout — APPROVED** [review]

**Review by kuisathaverat — APPROVED** [review]

**Pull request closed without merging**
